### PR TITLE
Reconcile by determining current state

### PR DESCRIPTION
**Pull request by sebbonnet**
This is a follow up to #150 and supercedes #164
This changes the reconciliation logic to reconstruct the Cassandra definition from the current state instead of tracking the previous resources state in memory which doesn't survive operator redeployments.

This approach allows for resources to be reconciled when the operator is restarted and _eventually_ when the resources are modified outside of the control of the operator (e.g. editing a statefulset, unintentionally deleted a job). I say _eventually_ because at the moment each change is applied asynchronously via our `dispatcher`. As a result, changes that fail the first time may not be applied before the next reconciliation loop (either at next change or at worst at the next `manager.Options.SyncPeriod` which defaults to 10hrs). Follow-up work will be done to address that, perhaps by swapping event dispatching and reconciliation, reducing the `SyncPeriod`, improving the changes application or a combination of that. 

As part of this change, the list of clusters is now managed by the `reconciler`
The list of clusters is mainly used to allow retrieval of metrics from
clusters managed by the operator.
Previously new clusters where only added on creation and on updates
which did not work for scenarios where the operator was restarted
but nothing needed to be reconciled.
This changes the logic to register new cluster definition at each
reconciliation, so the list is correct even when nothing is changed.

Required by #53

### CODE REVIEW

File: cassandra-operator/pkg/cluster/state.go [hunk 1]
```diff
@@ -0,0 +1,410 @@
+package cluster
```
Comment (alangibson01) [inline cassandra-operator/pkg/cluster/state.go]: I'm finding it hard to follow the content of what's in this file. Could we split it up into a couple of files, each with a couple of structs or interfaces in it? Some extra comments on what each of the types is supposed to do would be helpful as well.

File: cassandra-operator/test/e2e/pod_events.go [hunk 2]
```diff
@@ -57,24 +57,39 @@ func watchEvents(watcher watch.Interface, events eventLog) {
...
-		lastCreationTimeForFirstPod, err := e.lastCreationTimeForPod(firstPod)
-		if err != nil {
-			return false, err
-		}
-
-		deletionTimeForSecondPod, err := e.deletionTimeForPod(secondPod)
-		if err != nil {
-			return false, err
-		}
+func (e *PodEventLog) PodsRecreatedOneAfterTheOther(pod, otherPod string) (bool, error) {
```
Comment (alangibson01) [inline cassandra-operator/test/e2e/pod_events.go]: I think the naming of the arguments needs to be clearer. Which of the pods is the one we expect to be recreated first?

File: cassandra-operator/test/e2e/pod_events.go [hunk 3]
```diff
@@ -57,24 +57,39 @@ func watchEvents(watcher watch.Interface, events eventLog) {
...
+	}
+	deletionTimeForOtherPod, err := e.deletionTimeForPod(otherPod)
+	if err != nil {
+		return false, err
+	}
 
-		if deletionTimeForSecondPod.Before(lastCreationTimeForFirstPod) {
-			return false, fmt.Errorf("second pod was deleted before first pod was recreated. Second pod deleted at: %v, first pod last created at: %v", deletionTimeForSecondPod, lastCreationTimeForFirstPod)
-		}
+	if deletionTimeForPod.Before(deletionTimeForOtherPod) && !lastCreationTimeForPod.Before(deletionTimeForOtherPod) {
```
Comment (alangibson01) [inline cassandra-operator/test/e2e/pod_events.go]: I don't understand the logic of this check and the following one. Can we talk through it?

File: cassandra-operator/cmd/operator/controller.go [hunk 4]
```diff
@@ -17,118 +19,212 @@ import (
...
-		r.eventDispatcher.Dispatch(&dispatcher.Event{Kind: operations.DeleteCluster, Key: clusterID, Data: deletedCass})
-		delete(r.previousCassandras, request.NamespacedName.String())
+		logger.Debug("Cassandra definition not found. Going to delete associated resources")
+		cassandraToDelete := &v1alpha1.Cassandra{ObjectMeta: metav1.ObjectMeta{Name: request.Name, Namespace: request.Namespace}}
+		r.eventDispatcher.Dispatch(&dispatcher.Event{Kind: operations.DeleteCluster, Key: ctx.eventKey, Data: cassandraToDelete})
+		delete(r.clusters, ctx.namespaceName)
 		return reconcile.Result{}, nil
 	}
+	if err != nil {
+		logger.Errorf("Could not fetch Cassandra definition. Will retry. Error: %v", err)
```
[the remark this answers is not in the document]
Comment (sebbonnet) [reply cassandra-operator/cmd/operator/controller.go]: Yes - you're right, at the moment we get duplicate errors in the logs. I'll change that

File: cassandra-operator/cmd/operator/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+	AfterEach(func() {
+		fakes.assertAll(GinkgoT())
+	})
+
+	Describe("event dispatching", func() {
+
+		Context("cluster changes", func() {
+			It("should dispatch a DeleteCluster event when no Cassandra definition found", func() {
+				fakes.cassandraIsNotFoundIn(clusterNamespaceName)
+				fakes.dispatcher.On("Dispatch", &dispatcher.Event{
```
[the remark this answers is not in the document]
Comment (sebbonnet) [reply cassandra-operator/cmd/operator/controller_test.go]: sounds good

File: cassandra-operator/cmd/operator/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+			BeforeEach(func() {
+				configMap = &corev1.ConfigMap{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      desiredCassandra.CustomConfigMapName(),
+						Namespace: desiredCassandra.Namespace,
+					},
+				}
+			})
+
+			It("should dispatch configMap events before Cassandra ones", func() {
```
[the remark this answers is not in the document]
Comment (sebbonnet) [reply cassandra-operator/cmd/operator/controller_test.go]: it doesn't. there are no built-in way in testify and no appetite for either: https://github.com/stretchr/testify/issues/684

[the remark this answers is not in the document]
Comment (sebbonnet) [reply cassandra-operator/cmd/operator/controller_test.go]: we will need to build something specifically for this test to assert the order.

File: cassandra-operator/cmd/operator/main.go [hunk 7]
```diff
@@ -142,17 +144,10 @@ func startOperator(_ *cobra.Command, _ []string) error {
...
-			eventDispatcher:    dispatcher.New(receiver.Receive, stopCh),
-			eventRecorder:      eventRecorder,
-			previousCassandras: map[string]*v1alpha1.Cassandra{},
-			previousConfigMaps: map[string]*corev1.ConfigMap{},
-		},
+		Reconciler: NewReconciler(clusters, mgr.GetClient(), eventRecorder, dispatcher.New(receiver.Receive, stopCh)),
 	})
 	if err != nil {
-		entryLog.Error(err, "unable to set up individual controller")
+		entryLog.Error(err, "unable to set up Cassandra reconciler controller")
```
[the remark this answers is not in the document]
Comment (sebbonnet) [reply cassandra-operator/cmd/operator/main.go]: nope, I'll use Fatal

File: cassandra-operator/test/e2e/pod_events.go (hunk 2 above)
Comment (sebbonnet) [reply cassandra-operator/test/e2e/pod_events.go]: The order of the pods doesn't matter.
Whichever pod is deleted first has to be started before the other pod is deleted to ensure there is no downtime

File: cassandra-operator/test/e2e/pod_events.go (hunk 3 above)
Comment (sebbonnet) [reply cassandra-operator/test/e2e/pod_events.go]: The basic idea is to validate no 2 pods were down at the same time.
It doesn't matter which pod was first deleted.
If a pod was deleted before the other pod, it should have been (re)created before the other pod is deleted. If that's not the case then 2 pods were down at one moment in time, which is the error case.

Comment (alangibson01) [reply cassandra-operator/test/e2e/pod_events.go]: OK, that makes sense now. Thanks! Maybe we could rename the function to `PodsNotDownAtTheSameTime` ...?

File: cassandra-operator/pkg/cluster/state.go (hunk 1 above)
Comment (sebbonnet) [reply cassandra-operator/pkg/cluster/state.go]: I've now split state (and its test) into 3 different files that contain both the interface and its implementation:
- state.go exposing the exported interface
- clusterstate.go gathers code related to cluster (statefulset) reconstruction
- snapshot.go gathers code related to snapshot reconstruction

Also factory.go contains the objectRefenceFactory interface and its implementation

File: cassandra-operator/pkg/cluster/clusterstate.go [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+	sort.SliceStable(statefulSets.Items, func(i, j int) bool {
+		return statefulSets.Items[i].Name < statefulSets.Items[j].Name
+	})
+	return statefulSets, nil
+}
+
+func (cass *currentClusterStateFinder) buildPodSpecFrom(statefulSet *v1beta2.StatefulSet, desiredCassandra *v1alpha1.Cassandra) (v1alpha1.Pod, error) {
+	cassandraContainer, err := cass.containerWithName(cassandraContainerName, statefulSet)
+	if err != nil {
+		return v1alpha1.Pod{}, err
```
Comment (alangibson01) [inline cassandra-operator/pkg/cluster/clusterstate.go]: Any reason we don't change the return type to `*v1alpha1.Pod, error` and return `nil, err` here?

File: cassandra-operator/pkg/cluster/state_test.go [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+
+		AfterEach(func() {
+			mockInternalStateFinder.AssertExpectations(GinkgoT())
+		})
+
+		Context("an error occurs when finding the state", func() {
+
+			It("should return an error when unable to find the cluster state", func() {
+				unknownError := fmt.Errorf("unknown")
+				mockInternalStateFinder.On("findClusterStateFor", desiredCassandra).
```
Comment (alangibson01) [inline cassandra-operator/pkg/cluster/state_test.go]: Could we do something similar to what you've done in the other test files, and hide these calls underneath methods that describe the test state that's being set up?

File: cassandra-operator/pkg/cluster/clusterstate.go (hunk 8 above)
Comment (sebbonnet) [reply cassandra-operator/pkg/cluster/clusterstate.go]: good idea

File: cassandra-operator/pkg/cluster/state_test.go (hunk 9 above)
Comment (sebbonnet) [reply cassandra-operator/pkg/cluster/state_test.go]: seems like I got lazy :)
but let me add wrapper methods here too
